### PR TITLE
Distinguish between path_find responses (RIPD-1013)

### PR DESCRIPTION
**Pull request by miguelportilla**
Distinguish between the path_find subcommand responses by including new fields.

Reviewers: @rec @scottschurr

### CODE REVIEW

**Comment by scottschurr** [comment]
From a path finding novice this looks good to me.  :+1:.

I've asked @sublimator if this looks good, since he can represent the code that receives the JSON.

**Comment by sublimator** [comment]
I'm a little bit lost on what the problem is this is solving tbh. I will need to look into it more.

**Comment by scottschurr** [comment]
@sublimator, I just checked and apparently the request came from Chris Clark?  See https://ripplelabs.atlassian.net/browse/RIPD-1013.

**Comment by sublimator** [comment]
@scottschurr Thanks. 

I've been dimly aware of it, but haven't gotten around to fully understanding the problem.  Is the issue that no matter what `id` is passed in, that the subcommand responses will always use the id used initially with path_find/create ?

**Comment by miguelportilla** [comment]
@sublimator The issue is that there is no way to pair a path_find subcommand with a response. In other words a caller can't tell if the response they just received belongs to the create, close or status commands. IDs have nothing to do with the problem. I've explain how the id works and elaborated in the task thread. https://ripplelabs.atlassian.net/browse/RIPD-1013

**Comment by rec** [comment]
I think this is a good, backwards compatible solution to the bug.  

Tiny quibble: "Differentiate" confused me when I first saw it - I thought it was mathematics.

"Distinguish between" is unambiguous and has the same meaning in this context.

**Comment by rec** [comment]
(Oh, and the issue of "knowing which response belongs to which request" needs to be properly handled by rippled in a more systematic manner.  BUT that's a can of worms, this fixes the reported issue elegantly and with minimal delta...)

**Comment by sublimator** [comment]
The way ripple-lib works, because websockets are async (responses may not arrive in the same order as the requests)  is that you send a request, and give it an id, and if you set some callback to handle the reply, the way the reply is routed to that callback is by looking at the id in the messages, looking up the request object by id and notifying subscribers. 

Here's some json from rippled (without this patch):

## path_find/close
```json
 {
  "command": "path_find",
  "id": 52,
  "subcommand": "close"
}
```

## path_find/close response
```json
 {
  "id": 52,
  "result": {
    "alternatives": [],
    "destination_account": "rGpeQzUWFu4fMhJHZ1Via5aqFC3A5twZUD",
    "destination_amount": "10000000",
    "full_reply": true,
    "id": 51,
    "source_account": "r4nmQNH4Fhjfh6cHDbvVSsBv7KySbj4cBf"
  },
  "status": "success",
  "type": "response"
}
```

Note there's `result.id`  and `id` with the former being the id of the initial path_find/create. This is a websocket only api, so use of `id` in these requests is not really optional in practice. So the way you'd know which subcommand the response is for is, like any other request, by id. 

From looking at the code, it looks like `jvStatus` is a member variable that is mutated at various times, but never actually cleared, hence why above you can see what is essentially previous updates. 

From my reading the code will update the messages as so:

## path_find/close

```json
 {
  "id": 52,
  "result": {
    "closed": true,

    "alternatives": [],
    "destination_account": "rGpeQzUWFu4fMhJHZ1Via5aqFC3A5twZUD",
    "destination_amount": "10000000",
    "full_reply": true,
    "id": 51,
    "source_account": "r4nmQNH4Fhjfh6cHDbvVSsBv7KySbj4cBf"
  },
  "status": "success",
  "type": "response"
}
```

It seems like you'd want just the `id` and `closed` in the result ? Why is `closed` set to true while `status` is set to `success` ?

## path_find/status

```json
 {
  "id": 52,
  "result": {
    "status": "success",

    "alternatives": [],
    "destination_account": "rGpeQzUWFu4fMhJHZ1Via5aqFC3A5twZUD",
    "destination_amount": "10000000",
    "full_reply": true,
    "id": 51,
    "source_account": "r4nmQNH4Fhjfh6cHDbvVSsBv7KySbj4cBf"
  },
  "status": "success",
  "type": "response"
}
```

It seems like `result.status` is redundant there, given the outer framing. What is the purpose of the status subcommand?  I've never actually used it.

File: src/ripple/app/paths/PathRequest.cpp [hunk 1]
```diff
@@ -393,12 +393,14 @@ Json::Value PathRequest::doClose (Json::Value const&)
 {
     m_journal.debug << iIdentifier << " closed";
     ScopedLockType sl (mLock);
+    jvStatus[jss::closed] = true;
     return jvStatus;
 }
 
 Json::Value PathRequest::doStatus (Json::Value const&)
 {
     ScopedLockType sl (mLock);
```
Comment (sublimator) [inline src/ripple/app/paths/PathRequest.cpp]: So it acquires a lock, returns a  copy, then releases the lock?

File: src/ripple/app/paths/PathRequest.cpp [hunk 2]
```diff
@@ -393,12 +393,14 @@ Json::Value PathRequest::doClose (Json::Value const&)
 {
     m_journal.debug << iIdentifier << " closed";
     ScopedLockType sl (mLock);
+    jvStatus[jss::closed] = true;
```
Comment (sublimator) [inline src/ripple/app/paths/PathRequest.cpp]: Why not `"subcommand": "close"` ?

Comment (sublimator) [inline src/ripple/app/paths/PathRequest.cpp]: nvm, I guess that makes sense

**Comment by sublimator** [comment]
@scottschurr  @miguelportilla 

Looks ok to me, but I'm not sure I really get the issue

**Comment by scottschurr** [comment]
@sublimator Thanks for looking.  If you're content I'm content.

**Comment by rec** [comment]
Just making sure that this is still :+1: from me.

**Comment by scottschurr** [comment]
Incorporated as 464410d8be54945ff4539751ea2aa07d8dedf0f5.

**Pull request closed without merging**